core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * The container will go through all the registered Performance Counters and will trigger their 'report' method.
+ * By default the container will start reporting after 5 minutes and will continue doing so every 1 minute.
+ *
+ * The user of this class can add (register), remove (unregister) a performance counter while the container is working.
+ *
+ * The container will be stopped automatically when the application exists.
+ */
+public enum PerformanceCounterContainer implements Stoppable {
+    INSTANCE;
+
+    // By default the container will wait 2 minutes before the collection of performance data.
+    private final static long START_COLLECTING_INTERVAL_IN_MILLIS = 120000;
+
+    // By default the container will collect performance data every 1 minute.
+    private final static long COLLECTING_INTERVAL_IN_MILLIS = 60000;
+
+    private final ConcurrentMap<String, PerformanceCounter> performanceCounters = new ConcurrentHashMap<String, PerformanceCounter>();
+
+    private volatile boolean initialized = false;
+
+    private long startCollectingIntervalInMillis = START_COLLECTING_INTERVAL_IN_MILLIS;
+    private long collectingIntervalInMillis = COLLECTING_INTERVAL_IN_MILLIS;
+
+    private TelemetryClient telemetryClient;
+
+    private ScheduledThreadPoolExecutor threads;
+
+    /**
+     * Registers a {@link com.microsoft.applicationinsights.internal.perfcounter.PerformanceCounter} that can collect data.
+     * @param performanceCounter The Performance Counter.
+     */
+    public void register(PerformanceCounter performanceCounter) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(performanceCounter.getId()), "performanceCounter's id should be non null, non empty value");
+
+        initialize();
+        performanceCounters.put(performanceCounter.getId(), performanceCounter);
+    }
+
+    /**
+     * Un-registers a performance counter.
+     * @param performanceCounter The Performance Counter.
+     */
+    public void unregister(PerformanceCounter performanceCounter) {
+        unregister(performanceCounter.getId());
+    }
+
+    /**
+     * Un-registers a performance counter by its id.
+     * @param id The Performance Counter's id.
+     */
+    public void unregister(String id) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(id), "id should be non null, non empty value");
+
+        performanceCounters.remove(id);
+    }
+
+    /**
+     * Sets the timeout to wait before the first reporting.
+     *
+     * The number must be a positive number
+     *
+     * Note that the method will be effective if called before the first call to the 'register' method.
+     * @param startCollectingIntervalInMillis Timeout to wait before the first collection of performance counters in milliseconds.
+     */
+    public void setStartCollectingIntervalInMillis(long startCollectingIntervalInMillis) {
+        if (startCollectingIntervalInMillis <= 0) {
```

Comment:
basically not, when the application starts, I think we need to let it initialize for two reasons: 1. we don't want to take too much time, 2. the performance counters in this stage are not telling too much